services/inc/logging.h

```diff
@@ -0,0 +1,369 @@
+
+        unsigned int value = 1;
+        LOG_PRINTF(INFO, "%08x", value);
+
+    LOG_DUMP(level, data, size) - encodes data in hex and writes resulting string to backend logger.
+
+        extern struct Foo foo;
+        LOG_DUMP(TRACE, &foo, sizeof(foo));
+
+    LOG_ENABLED(level) - checks whether specified logging level is enabled at run time.
+
+        if (LOG_ENABLED(TRACE)) {
+            // Do some heavy logging
+        }
+
+    LOG_MODULE_CATEGORY - defines logging category at module level. This macro is typically defined
+    via compiler arguments.
+
+    LOG_SOURCE_CATEGORY(name) - defines logging category for a source file (.c, .cpp). Source category
+    overrides module category.
+
+        #include "logging.h"
+        LOG_SOURCE_CATEGORY("foo.bar")
+
+    LOG_CATEGORY(name) - defines logging category specific to particular scope, such as function body
+    or class declaration. Scoped category overrides any global or currently visible category.
+
+        void baz() {
+            LOG_CATEGORY("foo.bar.baz");
+            ...
+        }
+
+    Following macros take category name as argument:
+
+    LOG_C(level, category, format, ...)
+    LOG_WRITE_C(level, category, data, size)
+    LOG_PRINT_C(level, category, str)
+    LOG_PRINTF_C(level, category, format, ...)
+    LOG_DUMP_C(level, category, data, size)
+    LOG_ENABLED_C(level, category)
+
+    Every logging macro has its debugging counterpart which is compiled only in debug builds:
+
+        LOG(INFO, "User name: %s", user);
+        LOG_DEBUG(INFO, "User password: %s", password);
+        LOG_DUMP(TRACE, public_key, sizeof(public_key));
+        LOG_DEBUG_DUMP(TRACE, private_key, sizeof(private_key));
+        ...
+
+    Following macros configure logging module at compile time:
+
+    LOG_INCLUDE_SOURCE_INFO - allows to include source info (file name, function name, line number)
+    into messages generated by the logging macros. By default, this macro is defined only for debug
+    builds (see config.h).
+
+    LOG_COMPILE_TIME_LEVEL - allows to strip any logging output that is below of certain logging level
+    at compile time. Default value is LOG_LEVEL_ALL meaning that no compile-time filtering is applied.
+
+    LOG_MAX_STRING_LENGTH - specifies maximum number of characters allowed for formatted strings.
+    This parameter affects log_message() and some other functions along with their wrapper macros.
+
+    LOG_DISABLE - disables logging entirely, turning all logging macros into no-op.
+*/
+
+#include <string.h>
+#include <stdint.h>
+#include <stdarg.h>
+#include <stddef.h>
+#include "panic.h"
+#include "config.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum LogLevel {
+    LOG_LEVEL_ALL = 1, // Log all messages
+    LOG_LEVEL_TRACE = 1,
+    LOG_LEVEL_INFO = 30,
+    LOG_LEVEL_WARN = 40,
+    LOG_LEVEL_ERROR = 50,
+    LOG_LEVEL_PANIC = 60,
+    LOG_LEVEL_NONE = 70, // Do not log any messages
```

Comment:
There's an implicit dependency between the log level integer definitions and log_level_name. Maybe add a note to review log_level_name if this enum is changed.